java/ql/test/query-tests/security/CWE-378/semmle/tests/Test.java

```diff
@@ -0,0 +1,233 @@
+        temp.delete();
+        temp.mkdir();
+        return temp;
+    }
+
+    static File safe10() throws IOException {
+        File temp = File.createTempFile("test", "directory", new File(System.getProperty("user.dir")));
+        temp.delete();
+        temp.mkdir();
+        return temp;
+    }
+
+    static File vulnerable3() throws IOException {
+        File temp = File.createTempFile(
+            "test", 
+            "directory", 
+            new File(System.getProperty("java.io.tmpdir")).getAbsoluteFile()
+        );
+        temp.delete();
+        temp.mkdir();
+        return temp;
+    }
+
+    static File safe11() throws IOException {
+        File temp = null;
+        if (temp == null) {
+            while (true) {
+                temp = File.createTempFile("test", "directory");
+                if (temp.delete() && temp.mkdir()) {
+                    break;
+                }
+            }
+        }
+        return temp;
+    }
+
+    File safe12temp;
+    File safe12() throws IOException {
+        if (safe12temp == null) {
+            while (true) {
+                safe12temp = File.createTempFile("test", "directory");
+                if (safe12temp.delete() && safe12temp.mkdir()) {
+                    break;
+                }
+            }
+        }
+        return safe12temp;
+    }
+
+    File safe13() throws IOException {
+        File temp = File.createTempFile("test", "directory");
+        temp.delete();
+        if (temp.mkdir()) {
+            return temp;
+        } else {
+            throw new IOException(temp.getAbsolutePath() + " could not be created");
+        }
+    }
+
+    File safe14() throws IOException {
+        File temp = File.createTempFile("test", "directory");
+        temp.delete();
+        if (temp.mkdir()) {
+            return temp;
+        } else {
+            throw new IOException(temp.getAbsolutePath());
+        }
+    }
+
+    File safe15() throws IOException {
+        File temp = File.createTempFile("test", "directory");
+        temp.delete();
+        if (temp.mkdir()) {
+            return temp;
+        } else {
+            final String absolutePath = temp.getAbsolutePath();
+            throw new IOException(absolutePath);
+        }
+    }
+
+    File vulnerable4() throws IOException {
+        File temp = new File(System.getProperty("java.io.tmpdir"));
+        ensureDirectory(temp);
+        File workDir = File.createTempFile("test", "directory", temp);
+        if (!workDir.delete()) {
+            throw new IOException("Could not delete temp file: " + workDir.getAbsolutePath());
+        }
+        ensureDirectory(workDir);
+        return temp;
+    }
+    
+    private static void ensureDirectory(File dir) throws IOException {
+        if (!dir.mkdirs() && !dir.isDirectory()) {
+            throw new IOException("Mkdirs failed to create " + dir.toString());
+        }
+    }
```

Comment:
Before, when I wasn't using using `FlowState`, this actually flagged two vulnerabilities instead of one. The first one for the call to `ensureDirectory(temp);` and the second for the call to `ensureDirectory(workDir);`. Because I'm now using flow labels, it correctly, identifies the single path that is actually a vulnerability here.